src/cart/crt_iv.c

```diff
@@ -2564,6 +2567,75 @@ crt_ivu_rpc_issue(d_rank_t dest_rank, crt_iv_key_t *iv_key,
 	return rc;
 }
 
+static void
+handle_response_internal(void *arg)
+{
+	const struct crt_cb_info *cb_info = arg;
+	crt_rpc_t		 *rpc = cb_info->cci_rpc;
+
+	switch(rpc->cr_opc) {
+	case CRT_OPC_IV_FETCH:
+		handle_ivfetch_response(cb_info);
+		break;
+	case CRT_OPC_IV_SYNC:
+		handle_ivsync_response(cb_info);
+		break;
+	case CRT_OPC_IV_UPDATE:
+		handle_ivupdate_response(cb_info);
+		break;
+	default:
+		D_ERROR("wrong opc 0x%x\n", rpc->cr_opc);
+	}
+}
+
+static void
+handle_response_cb_internal(void *arg)
+{
+	struct crt_cb_info *cb_info = arg;
+	crt_rpc_t		 *rpc = cb_info->cci_rpc;
+	struct crt_rpc_priv	 *rpc_priv;
+
+	handle_response_internal(arg);
+
+	rpc_priv = container_of(rpc, struct crt_rpc_priv, crp_pub);
+	RPC_DECREF(rpc_priv);
+	D_FREE_PTR(cb_info);	
+}
+
+/* Fetch response handler */
+static void
+handle_response_cb(const struct crt_cb_info *cb_info)
+{
+	crt_rpc_t			*rpc = cb_info->cci_rpc;
+	struct crt_rpc_priv		*rpc_priv;
+	struct crt_context		*crt_ctx;
+
+	rpc_priv = container_of(rpc, struct crt_rpc_priv, crp_pub);
+	D_ASSERT(rpc_priv != NULL);
+	crt_ctx = rpc_priv->crp_pub.cr_ctx;
+
+	if (crt_rpc_cb_customized(crt_ctx, &rpc_priv->crp_pub)) {
+		int rc;
+		struct crt_cb_info *info;
+
+		D_ALLOC_PTR(info);
+		D_ASSERT(info != NULL);
+		RPC_ADDREF(rpc_priv);
+
+		info->cci_rpc = cb_info->cci_rpc;
+		info->cci_rc = cb_info->cci_rc;
+		info->cci_arg = cb_info->cci_arg;
+
+		rc = crt_ctx->cc_rpc_cb((crt_context_t)crt_ctx,
+					 (void *)info,
+					 handle_response_cb_internal,
+					 crt_ctx->cc_rpc_cb_arg);
+		D_ASSERT(rc == 0);
```

Comment:
Does daos ever return non-0 rc from cc_rpc_cb? if so, we should handle it gracefully

Reply:
yes, but I I would suggest to assert due to the reason above.

Reply:
I think if error is returned here we should at least make a call to regular handle_response_internal(cb_info); to let IV finalize its operation. The same could also be done if mem allocation above fails

Reply:
I think we should differentiate IV RPC and regular RPC, for example, allocate cb_info while creating the RPC and reduce the chance of failure during the RPC handling process at least.

Reply:
ah, both suggestions looks good to me.  For now, I will update the patch by Alex's idea, since it seems easier for the moment.